src/ring_theory/graded_algebra/homogeneous_ideal.lean

```diff
@@ -0,0 +1,335 @@
+iff.rfl
+
+lemma is_homogeneous_ideal_iff_subset_Inter (I : ideal R) :
+  is_homogeneous_ideal A I ↔ (I : set R) ⊆ ⋂ i, graded_algebra.proj A i ⁻¹' ↑I :=
+subset_Inter_iff.symm
+
+lemma is_homogeneous_ideal.exists_iff_eq_span :
+  (∃ (S : set (homogeneous_submonoid A)), I = ideal.span (coe '' S)) ↔
+  I = ideal.span {x | x ∈ I ∧ is_homogeneous A x} :=
+⟨λ ⟨S, hI⟩, begin
+    ext r, split; intro hr,
+  { rw hI at hr,
+    suffices : coe '' S ⊆ {x | x ∈ I ∧ is_homogeneous A x},
+    exact (ideal.span_mono this) hr,
+    intros s hs, split, rw hI,
+    refine ideal.subset_span hs,
+    obtain ⟨⟨s', homs'⟩, hs₁, hs₂⟩ := hs,
+    convert homs', rw ←hs₂, refl },
+  { obtain ⟨l, hl⟩ := (finsupp.mem_span_iff_total R _ _).mp hr,
+    rw ←hl, apply ideal.sum_mem, rintros ⟨x, hx₁, hx₂⟩ hx₃,
+    simp only [linear_map.id_coe, id.def, finsupp.mem_support_iff, linear_map.coe_smul_right,
+      ne.def, smul_eq_mul, subtype.coe_mk] at hx₁ hx₂ hx₃ ⊢,
+    exact ideal.mul_mem_left _ _ hx₁, }
+  end, λ hI, ⟨{x : homogeneous_submonoid A | ↑x ∈ I}, begin
+    rw hI, congr, ext r, split; intros hr,
+    { rcases hr with ⟨r_mem, ⟨i, r_eq⟩⟩,
+      use r, exact ⟨i, r_eq⟩, refine ⟨_, rfl⟩,
+      simp only [mem_set_of_eq, subtype.coe_mk], convert ←r_mem, },
+    { rcases hr with ⟨⟨r', hr'⟩, hr₁, hr₂⟩,
+      simp only [mem_set_of_eq, subtype.coe_mk] at hr₁,
+      rw ←hr₂, rw ←hI at hr₁, refine ⟨hr₁, hr'⟩, }
+  end⟩⟩
+
+variable [Π (i : ι) (x : A i), decidable (x ≠ 0)]
+
+lemma is_homogeneous_ideal.mul_homogeneous_element
+  {I : ideal R} (r x : R) (hx₁ : is_homogeneous A x) (hx₂ : x ∈ I) (j : ι) :
+  graded_algebra.proj A j (r * x) ∈ I :=
+begin
+  rw [←graded_algebra.sum_support_decompose A r, finset.sum_mul, linear_map.map_sum],
+  apply ideal.sum_mem,
+  intros k hk,
+  obtain ⟨i, hi⟩ := hx₁,
+  have mem₁ : (graded_algebra.proj A k) r * x ∈ A (k + i) := graded_monoid.mul_mem
+   (by { rw [graded_algebra.proj_apply], exact submodule.coe_mem _, }) hi,
+  by_cases k + i = j,
+  { rw ←h, rw graded_algebra.proj_apply at ⊢ mem₁,
+    rw graded_algebra.decompose_of_mem A mem₁, simp only [of_eq_same, submodule.coe_mk],
+    apply I.mul_mem_left _ hx₂, },
+  { rw [graded_algebra.proj_apply, graded_algebra.decompose_of_mem_ne],
+    exact I.zero_mem, exact mem₁, intro rid, apply h rid, }
+end
+
+lemma is_homogeneous_ideal.homogeneous_component
+  {I : ideal R} (HI : ∃ (S : set (homogeneous_submonoid A)), I = ideal.span (coe '' S))
+  (x : R) (hx : x ∈ I) (i : ι) :
+  graded_algebra.proj A i x ∈ I :=
+begin
+  have HI' := HI,
+  rw [is_homogeneous_ideal.exists_iff_eq_span, ideal.span,
+      finsupp.span_eq_range_total] at HI',
+  rw HI' at hx,
+  obtain ⟨s, rfl⟩ := hx,
+  rw [finsupp.total_apply, finsupp.sum, linear_map.map_sum],
+  refine ideal.sum_mem I _,
+  rintros ⟨j, ⟨hj₁, hj₂⟩⟩ hj₃,
+  simp only [algebra.id.smul_eq_mul, subtype.coe_mk, smul_eq_mul],
+  apply is_homogeneous_ideal.mul_homogeneous_element, exact hj₂, exact hj₁,
+end
```

Comment:
I think it would be best to inline this proof into the one below, the statement looks very strange otherwise.